fuzzers/libafl-fuzz/src/fuzzer.rs

```diff
@@ -0,0 +1,395 @@
+            testcase
+                .borrow_mut()
+                .add_metadata(IsInitialCorpusEntryMetadata {});
+        }
+    }
+
+    // Add the tokens to State
+    state.add_metadata(tokens);
+
+    // Set the start time of our Fuzzer
+    *state.start_time_mut() = current_time();
+
+    // Tell [`SeedFeedback`] that we're done loading seeds; rendering it benign.
+    fuzzer.feedback_mut().done_loading_seeds();
+
+    // Create a AFLStatsStage; TODO builder?
+    let afl_stats_stage = AflStatsStage::new(opt, fuzzer_dir.clone());
+
+    // Set LD_PRELOAD (Linux) && DYLD_INSERT_LIBRARIES (OSX) for target.
+    if let Some(preload_env) = &opt.afl_preload {
+        std::env::set_var("LD_PRELOAD", preload_env);
+        std::env::set_var("DYLD_INSERT_LIBRARIES", preload_env);
+    }
+
+    // Create a CmpLog executor if configured.
+    if let Some(ref cmplog_binary) = opt.cmplog_binary {
+        // The CmpLog map shared between the CmpLog observer and CmpLog executor
+        let mut cmplog_shmem = shmem_provider.uninit_on_shmem::<AFLppCmpLogMap>().unwrap();
+
+        // Let the Forkserver know the CmpLog shared memory map ID.
+        cmplog_shmem.write_to_env("__AFL_CMPLOG_SHM_ID").unwrap();
+        let cmpmap = unsafe { OwnedRefMut::from_shmem(&mut cmplog_shmem) };
+
+        // Create the CmpLog observer.
+        let cmplog_observer = AFLppCmpLogObserver::new("cmplog", cmpmap, true);
+        let cmplog_ref = cmplog_observer.handle();
+
+        // Create the CmpLog executor.
+        // Cmplog has 25% execution overhead so we give it double the timeout
+        let cmplog_executor = base_executor(opt, &mut shmem_provider)
+            .timeout(Duration::from_millis(opt.hang_timeout * 2))
+            .program(cmplog_binary)
+            .build(tuple_list!(cmplog_observer))
+            .unwrap();
+
+        // Create the CmpLog tracing stage.
+        let tracing = AFLppCmplogTracingStage::new(cmplog_executor, cmplog_ref);
+
+        // Create a randomic Input2State stage
+        let rq = MultiMutationalStage::new(AFLppRedQueen::with_cmplog_options(true, true));
+
+        // Create an IfStage and wrap the CmpLog stages in it.
+        // We run cmplog on the second fuzz run of the testcase.
+        // This stage checks if the testcase has been fuzzed more than twice, if so do not run cmplog.
+        // We also check if it is an initial corpus testcase
+        // and if run with AFL_CMPLOG_ONLY_NEW, then we avoid cmplog.
+        let cb = |_fuzzer: &mut _,
+                  _executor: &mut _,
+                  state: &mut LibaflFuzzState,
+                  _event_manager: &mut _|
+         -> Result<bool, Error> {
+            let testcase = state.current_testcase()?;
+            if opt.cmplog_only_new && testcase.has_metadata::<IsInitialCorpusEntryMetadata>() {
+                return Ok(false);
+            }
+            Ok(testcase.scheduled_count() == 1)
```

Comment:
Check the scheduled_count first, this check is basically free, while `has_metadata` is a super costly hashmap lookup